packages/core/src/rules/common/__tests__/no-required-schema-properties-undefined.test.ts

```diff
@@ -0,0 +1,391 @@
+          openapi: 3.0.0
+          components:
+            schemas:
+              Cat:
+                description: A representation of a cat
+                allOf:
+                  - $ref: '#/components/schemas/Pet'
+                  - type: object
+                    properties:
+                      huntingSkill:
+                        type: string
+                        description: The measured skill for hunting
+                        default: lazy
+                        example: adventurous
+                        enum:
+                          - clueless
+                          - lazy
+                          - adventurous
+                          - aggressive
+                    required:
+                      - huntingSkill
+                      - name
+              Pet:
+                type: object
+                required:
+                  - photoUrls
+                properties:
+                  name:
+                    description: The name given to a pet
+                    type: string
+                    example: Guru
+                  photoUrls:
+                    description: The list of URL to a cute photos featuring pet
+                    type: array
+                    maxItems: 20
+                    xml:
+                      name: photoUrl
+                      wrapped: true
+                    items:
+                      type: string
+                      format: url
+        `,
+      'foobar.yaml'
+    );
+
+    const results = await lintDocument({
+      externalRefResolver: new BaseResolver(),
+      document,
+      config: await makeConfig({ 'no-required-schema-properties-undefined': 'error' }),
+    });
+
+    expect(replaceSourceWithRef(results)).toMatchInlineSnapshot(`
+      [
+        {
+          "location": [
+            {
+              "pointer": "#/components/schemas/Cat/allOf/1/required/1",
+              "reportOnKey": false,
+              "source": "foobar.yaml",
+            },
+          ],
+          "message": "Required property 'name' is undefined.",
+          "ruleId": "no-required-schema-properties-undefined",
+          "severity": "error",
+          "suggest": [],
+        },
+      ]
+    `);
+  });
+
+  it('should not report required properties are present after resolving $refs when used in schema with allOf keyword', async () => {
+    const document = parseYamlToDocument(
+      outdent`
+          openapi: 3.0.0
+          components:
+            schemas:
+              Cat:
+                description: A representation of a cat
+                allOf:
+                  - $ref: '#/components/schemas/Pet'
+                  - type: object
+                    properties:
+                      huntingSkill:
+                        type: string
+                        description: The measured skill for hunting
+                        default: lazy
+                        example: adventurous
+                        enum:
```

Comment:
just a nitpick, I would clean up those specs to only required minimum, like we don't need enum or example or default to test the allOf case.

Reply:
cleaned up!